Feature/person presentation viewmodel

skaik-mo opened the pull request:
**Note: Use cases are not yet used in ViewModels.**

skaik-mo committed:
update BaseViewModel and create CostDetailsEffect

- Added `completion` parameter to `tryToExecute` method in `BaseViewModel`.
- Created `CostDetailsEffect` sealed interface.
- Updated `CastDetailsViewModel` to inherit from the refactored `BaseViewModel` and use `CostDetailsEffect`.

skaik-mo committed:
Implement gallery feature with new components and view model

skaik-mo committed:
Merge branch 'develop' into feature/person-presentation-viewmodel

* develop:
  Refactor: Update test coverage configuration to use unified media module
  Refactor: Rename package from com.giraffe.review to com.giraffe.media in ExampleInstrumentedTest
  This commit consolidates the individual media data source modules (movie, series, person, review, explore) into two unified modules: `datasource:local:media` and `datasource:remote:media`.
  Refactor: Remove unused domain modules from build configuration
  Add just one media repository
  Refactor: Comment out domain media dependency in build configuration
  Refactor: Introduce new data classes and update build configuration for media repository
  clean up and optimize imports
  update dependency for domain media and correct package name & path
  join movies, series, actors in one feature media

HendSayed25 commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GallerySection.kt:
```diff
@@ -0,0 +1,60 @@
+            onClickableText = onShowMoreClick,
+            modifier = Modifier.fillMaxWidth()
+        )
+        GalleryItemLayoutLTR(images = images, imageContentDescriptions = imageContentDescriptions)
```
how choose between LTR and RTL ?

skaik-mo replied:
Design: If the gallery is LTR, we’ll use GalleryItemLayoutLTR; otherwise, we’ll apply the RTL.

HendSayed25 replied:
I mean you call LTR here only
how will you use it ?

skaik-mo replied:
yup

skaik-mo committed:
Merge branch 'develop' into feature/person-presentation-viewmodel

* develop:
  Refactor: Update exception handling by replacing NetworkException with NoInternetException
  refactor: simplify searchMovieByName method by removing nested SafeCall
  Refactor: Rename media explore package and update related classes
  Refactor: Consolidate media data source modules and update response DTOs
  refactor: unify error handling by consolidating exception classes and updating error messages

skaik-mo committed:
Merge remote-tracking branch 'origin/develop' into feature/person-presentation-viewmodel

skaik-mo committed:
Integrate IslamicAppropriateImageViewer into GalleryItem and enhance CastDetailsScreen

The `CastDetailsScreen` has been updated to:
- Display a loading indicator while fetching data.
- Fetch and display actor details (name, image, birth info, gallery, biography, and movie credits) from the `GetPersonDetailsUseCase`.
- Use localized strings for section titles like "Best of [Actor Name]" and "Biography".
- Sample data updates in `MoviesListSection` preview for better text overflow testing.
- Added new string resources for "Best of" and "Biography".

Additionally, the `GallerySection` and related layout components have been simplified by removing the `imageContentDescriptions` parameter as it's now handled within `GalleryItem`.

skaik-mo committed:
Pass image list to GalleryScreen and ViewModel

dev-bilal-azzam commented on app/src/main/java/com/giraffe/cineverseapp/MainActivity.kt:
```diff
@@ -16,10 +17,7 @@ class MainActivity : ComponentActivity() {
-                    modifier = Modifier
-                        .fillMaxSize()
-                )
+                CastDetailsScreen()
```
don't push test code

mohannadahmed00 commented on app/src/main/java/com/giraffe/cineverseapp/MainActivity.kt:
don't play here. let it as it is.

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/base/BaseViewModel.kt:
```diff
@@ -26,13 +27,16 @@ abstract class BaseViewModel<S, E>(initialState: S): ViewModel() {
         onSuccess: (T) -> Unit = {},
         coroutineScope: CoroutineScope = viewModelScope,
         dispatcher: CoroutineDispatcher = Dispatchers.IO,
+        completion: () -> Unit = {},
```
just for consistency name it onComplete()

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItem.kt:
```diff
@@ -1,34 +1,57 @@
     modifier: Modifier = Modifier
 ) {
-    if (image == null) {
+    val model = ImageRequest
```
why you need this, the islamic appropriate already handling and requesting the image??

dev-bilal-azzam replied:
in this way you are requesting the image twice!

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItem.kt:
```diff
@@ -1,34 +1,57 @@
 fun GalleryItem(
-    image: Painter?,
-    contentDescription: String?,
+    image: String?,
```
rename it into imageUrl.

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItemLayout.kt:
```diff
@@ -0,0 +1,102 @@
+            modifier = Modifier.weight(0.5f)
+
+        ) {
+            images.getOrNull(1)?.let {
```
why not: images.forEach { it.let{________}}

skaik-mo replied:
I’m using `Column` and `Row` layouts here, and in some cases, I only take 3 items from the list to display.

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItemLayout.kt:
```diff
@@ -0,0 +1,102 @@
+
+@Composable
+fun GalleryItemLayoutLTR(
+    images: List<String?>,
```
rename it into imagesUrls.

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt:
```diff
@@ -44,15 +46,36 @@ fun CastDetailsScreen(
     castDetailsViewModel: CastDetailsViewModel = koinViewModel()
```
add modifier to the screen to control it from calling composable

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItemLayout.kt:
```diff
@@ -0,0 +1,102 @@
+                GalleryItem(
+                    image = it,
+                    modifier = Modifier
+                        .fillMaxHeight(0.5f)
```
just use weight(1f) here and with the below item, fillMaxHeight(0.5f) not needed.

skaik-mo replied:
not working in my case

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt:
```diff
@@ -44,15 +46,36 @@ fun CastDetailsScreen(
+    } else {
+        CastDetailsContent(
+            state = state,
+            onYoutubeClick = castDetailsViewModel::onActorYoutubeLinkClicked,
```
isn't it better to pass actions: CastDetailsViewModel instead of sending actions solely

mohannadahmed00 replied:
use ScreenInteractions interface here, it gonna be easier.

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/components/gallery/GalleryItemLayout.kt:
```diff
@@ -0,0 +1,102 @@
+                image = it,
+                modifier = Modifier
+                    .weight(1f)
+                    .fillMaxSize()
```
just use fillMaxHeight(), weight(1f) will handle the width calculations.

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsViewModel.kt:
```diff
@@ -1,10 +1,20 @@
 
 class CastDetailsViewModel(
+    val getPersonDetailsUseCase: GetPersonDetailsUseCase
+) : BaseViewModel<CastDetailsUiState, CostDetailsEffect>(CastDetailsUiState()),
```
rename CostDetailsEffect to CastDetailsEffect

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsViewModel.kt:
```diff
@@ -29,4 +39,41 @@ class CastDetailsViewModel(
         TODO("Not yet implemented")
     }
 
+    private fun getPersonDetails(personId: Int = 1) = safeExecute(
```
remove default personId send it from the screen:
you have 2 approaches inject it in the viewmodel constructor or call getPersonDetails from LaunchedEffect in the screen.

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CostDetailsEffect.kt:
remove this interface if you don't use it

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsViewModel.kt:
```diff
@@ -29,4 +39,41 @@ class CastDetailsViewModel(
     }
 
+    private fun getPersonDetails(personId: Int = 1) = safeExecute(
+        block = {
```
i think it is better to call safeExcecute like this safeExcecute(::onError, ::onSuccess){___________}

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/GalleryEffect.kt:
remove this interface if you don't use it

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/GalleryUiState.kt:
```diff
@@ -0,0 +1,6 @@
+
+data class GalleryUiState (
+    val actorName: String = "",
+    val images: List<String> = emptyList()
```
rename it into imagesUrls

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsViewModel.kt:
```diff
@@ -29,4 +39,41 @@ class CastDetailsViewModel(
+
+    private fun getPersonDetailsError(exception: Throwable) {
+        updateState { it.copy(isLoading = false) }
+        Log.d("Error", "getPersonDetailsError: ${exception.message}")
```
remove this line!

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/GalleryViewModel.kt:
remove it if you don't use it

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CostDetailsEffect.kt:
```diff
@@ -0,0 +1,5 @@
+package com.giraffe.details.screens.castDetails
+
+sealed interface CostDetailsEffect {
```
you might add an Effect for Error.
maybe to show Toast, snackbar or anything else

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/GalleryScreen.kt:
```diff
@@ -0,0 +1,81 @@
+
+@Preview
+@Composable
+fun GalleryScreen(
```
add modifier

dev-bilal-azzam commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/GalleryViewModel.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.giraffe.details.base.BaseViewModel
+
+class GalleryViewModel(
```
why you need this view model

skaik-mo committed:
Rename completion parameter to onCompletion in BaseViewModel

skaik-mo committed:
Replace CastDetailsScreen with ExploreScreen in MainActivity

skaik-mo committed:
Refactor: Use `imageUrl` consistently in gallery and details components

The following changes were made:
- Renamed `image` to `imageUrl` in `GalleryItem`, `GallerySection`, `GalleryItemLayoutLTR`, and `GalleryItemLayoutRTL`.
- Updated `MainDetails` and `MainDetailsHeader` to accept `actorImageUrl` instead of `actorImage` (Painter).
- Modified `CastDetailsUiState` to use `actorImageUrl` and `actorGalleryImageUrls`.
- Adapted `CastDetailsScreen` and related previews to reflect these changes.
- Updated `GalleryScreen` and `GalleryUiState` to use `imageUrls`.
- Removed `GalleryEffect.kt` and `GalleryViewModel.kt` as they are no longer needed after the refactoring.

skaik-mo committed:
Refactor: Adjust layout modifiers in GalleryItemLayout for better responsiveness

skaik-mo committed:
Refactor: Simplify GalleryItem by removing unnecessary image loading logic

skaik-mo committed:
Merge branch 'develop' into feature/person-presentation-viewmodel

* develop:
  customize contentScale,alignment,alpha, colorFilter, filterQuality, transform
  rename to safe islamic
  fix image viewer
  fix image viewer
  fix image viewer
  fix image viewer
  add primary key tp person cache dto
  refactor media local data source packages and names
  refactor media local data source packages and names

skaik-mo committed:
Replaced `IslamicAppropriateImageViewer` with the more generic `SafeIslamicImage` component across `MainDetails.kt` and `GalleryItem.kt`.

- Improved the placeholder display in `MainDetails.kt` and `GalleryItem.kt` by:
    - Adding a border and background to the placeholder icon.
    - Using `wrapContentSize` for better sizing.
- Streamlined null checks in `GalleryItemLayout.kt` for image URLs.
- Adjusted modifiers for consistent sizing and styling of images and placeholders.

skaik-mo committed:
Refactor: Simplify CastDetailsContent by consolidating interaction callbacks

skaik-mo committed:
Refactor: Enhance CastDetailsViewModel and related components

- Rename `CostDetailsEffect` to `CastDetailsEffect` and add an `Error` state.
- Update `CastDetailsViewModel` to accept `personId` in the constructor and fetch person details accordingly.
- Implement error handling in `CastDetailsViewModel` by sending an `Error` effect.
- Update `CastDetailsScreen` to receive `personId` and handle `CastDetailsEffect.Error`.
- Make `GalleryScreen` accept a `modifier`.
- Correct key generation for shared element transitions in `MainDetails`.
